File: packages/dotnet/DeRec.Library/src/Flows/Sharing.cs
using System;
using System.Collections.Generic;

namespace DeRec.Library;

public static class Sharing
{
    /// <summary>
    /// Result of <see cref="ProtectSecret"/>.
    ///
    /// Contains one serialized <c>CommittedDeRecShare</c> per helper channel in the
    /// binary FFI container format (count + per-entry channel ID + length-prefixed bytes).
    ///
    /// Pass each entry to <c>ProduceStoreShareRequestMessage</c> together with the
    /// helper's shared key to produce the encrypted delivery envelope.
    /// </summary>
    public sealed class ProtectSecretResult
    {
        public required byte[] SharesWireBytes { get; init; }
    }

    /// <summary>
    /// Result of <see cref="ProduceStoreShareRequestMessage"/>.
    ///
    /// Contains a serialized <c>DeRecMessage</c> envelope whose inner payload is an
    /// encrypted <c>StoreShareRequestMessage</c>.
    ///
    /// Send <see cref="WireBytes"/> to the helper over the channel transport.
    /// The helper stores these bytes and uses them to respond to verification
    /// and recovery requests.
    /// </summary>
    public sealed class ProduceStoreShareRequestMessageResult
    {
        public required byte[] WireBytes { get; init; }
    }

    /// <summary>
    /// Result of <see cref="ProduceStoreShareResponseMessage"/>.
    ///
    /// Contains both the encrypted response to send back to the Owner and the
    /// committed share bytes for local storage.
    /// </summary>
    public sealed class ProduceStoreShareResponseMessageResult
    {
        /// <summary>
        /// Serialized <c>DeRecMessage</c> envelope carrying an encrypted
        /// <c>StoreShareResponseMessage</c>. Send these bytes back to the Owner.
        /// </summary>
        public required byte[] WireBytes { get; init; }

        /// <summary>
        /// Serialized <c>CommittedDeRecShare</c> protobuf bytes extracted from the request.
        /// The Helper must persist this value for future verification and recovery requests.
        /// </summary>
        public required byte[] CommittedShareBytes { get; init; }
    }

    /// <summary>
    /// Generates verifiable secret shares and returns one serialized
    /// <c>CommittedDeRecShare</c> per helper channel.
    /// </summary>
    /// <param name="secretId">Identifier of the secret being protected.</param>
    /// <param name="secretData">Raw secret bytes to split using VSS.</param>
    /// <param name="channelIds">Helper channel identifiers. Duplicate entries are deduplicated.</param>
    /// <param name="threshold">Minimum number of shares required for reconstruction.</param>
    /// <param name="version">Logical version of this secret distribution.</param>
    /// <returns>
    /// A <see cref="ProtectSecretResult"/> whose <c>SharesWireBytes</c> contain a
    /// serialized map from channel ID to <c>CommittedDeRecShare</c> bytes.
    /// </returns>
    public static ProtectSecretResult ProtectSecret(
        byte[] secretId,
        byte[] secretData,
        ulong[] channelIds,
        ulong threshold,
        int version
    )
    {
        if (channelIds is null)
        {
            throw new ArgumentNullException(nameof(channelIds));
        }

        Native.Sharing.ProtectSecretResult nativeResult =
            Native.Sharing.protect_secret(
                secretId,
                (UIntPtr)secretId.Length,
                secretData,
                (UIntPtr)secretData.Length,
                channelIds,
                (UIntPtr)channelIds.Length,
                (UIntPtr)threshold,
                version
            );

        try
        {
            Utils.ThrowIfError(nativeResult.Status);

            byte[] sharesWireBytes = Utils.CopyBuffer(nativeResult.SharesWireBytes);

            return new ProtectSecretResult
            {
                SharesWireBytes = sharesWireBytes,
            };
        }
        finally
        {
            Utils.FreeBuffer(nativeResult.SharesWireBytes);
            Utils.FreeStatusMessage(nativeResult.Status);
        }
    }

    /// <summary>
    /// Wraps a committed helper share into an encrypted delivery envelope.
    ///
    /// Call this once for each share entry returned by <see cref="ProtectSecret"/>,
    /// providing the corresponding helper's shared key (established during pairing).
    /// Send the resulting <see cref="ProduceStoreShareRequestMessageResult.WireBytes"/>
    /// to the helper over the channel transport.
    /// </summary>
    /// <param name="channelId">Channel ID of the target helper.</param>
    /// <param name="version">Share-distribution version number.</param>
    /// <param name="committedShare">
    /// Serialized <c>CommittedDeRecShare</c> protobuf bytes for this channel,
    /// taken from the deserialized map produced by <see cref="ProtectSecret"/>.
    /// </param>
    /// <param name="keepList">
    /// Version numbers the helper should retain. Pass an empty array to use the
    /// helper's default retention policy.
    /// </param>
    /// <param name="description">Human-readable description of this share distribution.</param>
    /// <param name="sharedKey">32-byte symmetric key shared with this helper.</param>
    /// <returns>
    /// A <see cref="ProduceStoreShareRequestMessageResult"/> whose
    /// <c>WireBytes</c> are a serialized <c>DeRecMessage</c> envelope.
    /// </returns>
    public static ProduceStoreShareRequestMessageResult ProduceStoreShareRequestMessage(
        ulong channelId,
        int version,
        byte[] committedShare,
        int[] keepList,
        string description,
        byte[] sharedKey
    )
    {
        if (committedShare is null)
        {
            throw new ArgumentNullException(nameof(committedShare));
        }

        if (keepList is null)
        {
            throw new ArgumentNullException(nameof(keepList));
        }

        if (sharedKey is null)
        {
            throw new ArgumentNullException(nameof(sharedKey));
        }

        if (sharedKey.Length != 32)
        {
            throw new ArgumentException("sharedKey must be exactly 32 bytes.", nameof(sharedKey));
        }

        byte[] descriptionBytes = System.Text.Encoding.UTF8.GetBytes(description ?? string.Empty);

        Native.Sharing.ProduceStoreShareRequestMessageResult nativeResult =
            Native.Sharing.produce_store_share_request_message(
                channelId,
                version,
                committedShare,
                (UIntPtr)committedShare.Length,
                keepList,
                (UIntPtr)keepList.Length,
                descriptionBytes,
                (UIntPtr)descriptionBytes.Length,
                sharedKey,
                (UIntPtr)sharedKey.Length
            );

        try
        {
            Utils.ThrowIfError(nativeResult.Status);

            return new ProduceStoreShareRequestMessageResult
            {
                WireBytes = Utils.CopyBuffer(nativeResult.WireBytes),
            };
        }
        finally
        {
            Utils.FreeBuffer(nativeResult.WireBytes);
            Utils.FreeStatusMessage(nativeResult.Status);
        }
    }

    /// <summary>
    /// Processes an incoming sharing request on behalf of a Helper.
    ///
    /// Decrypts and validates the <c>StoreShareRequestMessage</c> carried inside the
    /// provided <c>DeRecMessage</c> envelope, extracts the committed share, and returns
    /// an encrypted <c>StoreShareResponseMessage</c> plus the committed share bytes for
    /// local storage.
    /// </summary>
    /// <param name="channelId">Channel ID of the Owner channel this request arrived on.</param>
    /// <param name="sharedKey">32-byte symmetric key shared with the Owner.</param>
    /// <param name="requestBytes">
    /// Serialized <c>DeRecMessage</c> envelope bytes received from the Owner,
    /// as produced by <see cref="ProduceStoreShareRequestMessage"/>.
    /// </param>
    /// <returns>
    /// A <see cref="ProduceStoreShareResponseMessageResult"/> containing the response wire
    /// bytes and the serialized committed share bytes for local storage.
    /// </returns>
    public static ProduceStoreShareResponseMessageResult ProduceStoreShareResponseMessage(
        ulong channelId,
        byte[] sharedKey,
        byte[] requestBytes
    )
    {
        if (sharedKey is null)
        {
            throw new ArgumentNullException(nameof(sharedKey));
        }

        if (sharedKey.Length != 32)
        {
            throw new ArgumentException("sharedKey must be exactly 32 bytes.", nameof(sharedKey));
        }

        if (requestBytes is null)
        {
            throw new ArgumentNullException(nameof(requestBytes));
        }

        Native.Sharing.ProduceStoreShareResponseMessageResult nativeResult =
            Native.Sharing.produce_store_share_response_message(
                channelId,
                sharedKey,
                (UIntPtr)sharedKey.Length,
                requestBytes,
                (UIntPtr)requestBytes.Length
            );

        try
        {
            Utils.ThrowIfError(nativeResult.Status);

            return new ProduceStoreShareResponseMessageResult
            {
                WireBytes = Utils.CopyBuffer(nativeResult.WireBytes),
                CommittedShareBytes = Utils.CopyBuffer(nativeResult.CommittedShareBytes),
            };
        }
        finally
        {
            Utils.FreeBuffer(nativeResult.WireBytes);
            Utils.FreeBuffer(nativeResult.CommittedShareBytes);
            Utils.FreeStatusMessage(nativeResult.Status);
        }
    }

    /// <summary>
    /// Validates a sharing response received from a Helper.
    ///
    /// Decrypts and validates the <c>StoreShareResponseMessage</c> carried inside the
    /// provided <c>DeRecMessage</c> envelope. Checks the timestamp invariant, that the
    /// echoed version matches the supplied <paramref name="version"/>, and that the
    /// <c>result</c> field is present.
    /// </summary>
    /// <param name="version">Version number that was sent in the original request.</param>
    /// <param name="sharedKey">32-byte symmetric key shared with the Helper.</param>
    /// <param name="responseBytes">
    /// Serialized <c>DeRecMessage</c> envelope bytes received from the Helper,
    /// as produced by <see cref="ProduceStoreShareResponseMessage"/>.
    /// </param>
    /// <exception cref="DeRecException">
    /// Thrown if decryption, decoding, or invariant checks fail, or if the Helper's
    /// response status is not <c>Ok</c>. The exception message includes the Helper's
    /// status code and memo when the rejection originates from the Helper.
    /// </exception>
    public static void ProcessStoreShareResponseMessage(
        int version,
        byte[] sharedKey,
        byte[] responseBytes
    )
    {
        if (sharedKey is null)
        {
            throw new ArgumentNullException(nameof(sharedKey));
        }

        if (sharedKey.Length != 32)
        {
            throw new ArgumentException("sharedKey must be exactly 32 bytes.", nameof(sharedKey));
        }

        if (responseBytes is null)
        {
            throw new ArgumentNullException(nameof(responseBytes));
        }

        Native.Sharing.ProcessStoreShareResponseMessageResult nativeResult =
            Native.Sharing.process_store_share_response_message(
                version,
                sharedKey,
                (UIntPtr)sharedKey.Length,
                responseBytes,
                (UIntPtr)responseBytes.Length
            );

        try
        {
            Utils.ThrowIfError(nativeResult.Status);
        }
        finally
        {
            Utils.FreeStatusMessage(nativeResult.Status);
        }
    }
}

File: packages/go/primitives/sharing/sharing.go
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 DeRec Alliance. All rights reserved.

// Package sharing wraps the DeRec sharing flow: splitting a secret into
// per-channel committed shares and exchanging store-share request/response
// messages that persist them with helpers.
package sharing

import (
	"encoding/binary"
	"fmt"

	"github.com/derecalliance/lib-derec/packages/go/internal/native"
)

// ExtractedRequest is the decrypted store-share request: the channel it
// arrived on and the inner StoreShareRequestMessage proto bytes, which chain
// into response production.
type ExtractedRequest struct {
	ChannelID    uint64
	RequestProto []byte
}

// ProducedResponse is a produced store-share response: the wire-encoded
// envelope to send back, the committed share the helper persisted, and the
// secret id/version it was persisted under.
type ProducedResponse struct {
	Envelope       []byte
	CommittedShare []byte
	SecretID       uint64
	Version        uint32
}

// ExtractedResponse is the decrypted store-share response: the channel it
// arrived on and the inner StoreShareResponseMessage proto bytes, which chain
// into Response.Process.
type ExtractedResponse struct {
	ChannelID     uint64
	ResponseProto []byte
}

type requestAPI struct{}

// Request groups the owner-side sharing operations, mirroring the Rust
// primitives::sharing::request module.
var Request requestAPI

// Split shares secretData across channelIDs using a Shamir secret-sharing
// scheme at the given threshold, returning one committed share per channel
// keyed by channel id.
func (requestAPI) Split(secretID uint64, secretData []byte, channelIDs []uint64, threshold int, version uint32) (map[uint64][]byte, error) {
	wire, err := native.ProtectSecret(secretID, secretData, channelIDs, uintptr(threshold), version)
	if err != nil {
		return nil, err
	}
	return decodeShares(wire)
}

// Produce builds the wire-encoded DeRecMessage carrying committedShare for
// storage at channelID under secretID/version, encrypted under sharedKey.
func (requestAPI) Produce(channelID uint64, version uint32, secretID uint64, committedShare []byte, keepList []uint32, description string, sharedKey []byte) ([]byte, error) {
	return native.ProduceStoreShareRequest(channelID, version, secretID, committedShare, keepList, description, sharedKey)
}

// Extract decrypts a store-share request envelope and returns its channel id
// and inner proto bytes.
func (requestAPI) Extract(request, sharedKey []byte) (ExtractedRequest, error) {
	channelID, proto, err := native.ExtractStoreShareRequest(request, sharedKey)
	if err != nil {
		return ExtractedRequest{}, err
	}
	return ExtractedRequest{ChannelID: channelID, RequestProto: proto}, nil
}

type responseAPI struct{}

// Response groups the helper-side sharing operations, mirroring the Rust
// primitives::sharing::response module.
var Response responseAPI

// Produce builds the wire-encoded DeRecMessage acknowledging requestProto
// (the RequestProto returned by Request.Extract), encrypted under sharedKey,
// and returns the committed share the helper should persist for later
// recovery responses.
func (responseAPI) Produce(channelID uint64, requestProto, sharedKey []byte) (ProducedResponse, error) {
	wireBytes, committedShareBytes, secretID, version, err := native.ProduceStoreShareResponse(channelID, requestProto, sharedKey)
	if err != nil {
		return ProducedResponse{}, err
	}
	return ProducedResponse{
		Envelope:       wireBytes,
		CommittedShare: committedShareBytes,
		SecretID:       secretID,
		Version:        version,
	}, nil
}

// Extract decrypts a store-share response envelope and returns its channel id
// and inner proto bytes.
func (responseAPI) Extract(response, sharedKey []byte) (ExtractedResponse, error) {
	channelID, proto, err := native.ExtractStoreShareResponse(response, sharedKey)
	if err != nil {
		return ExtractedResponse{}, err
	}
	return ExtractedResponse{ChannelID: channelID, ResponseProto: proto}, nil
}

// Process validates a store-share response (the ResponseProto returned by
// Response.Extract) against the expected version.
func (responseAPI) Process(version uint32, responseProto []byte) error {
	return native.ProcessStoreShareResponse(version, responseProto)
}

// decodeShares unpacks the committed-shares wire format documented at
// library/src/interop/ffi/sharing.rs into a channel id -> serialized
// CommittedDeRecShare map:
//
//	[count: u32 LE]
//	for each entry (sorted by channel ID):
//	  [channel_id: u64 LE]
//	  [share_len: u32 LE]
//	  [serialized CommittedDeRecShare protobuf]
func decodeShares(wire []byte) (map[uint64][]byte, error) {
	if len(wire) < 4 {
		return nil, fmt.Errorf("sharing: shares wire bytes too short for count: got %d bytes", len(wire))
	}
	count := binary.LittleEndian.Uint32(wire)
	offset := 4

	out := make(map[uint64][]byte, count)
	for i := uint32(0); i < count; i++ {
		if offset+8+4 > len(wire) {
			return nil, fmt.Errorf("sharing: unexpected end of shares wire bytes reading entry %d header", i)
		}
		channelID := binary.LittleEndian.Uint64(wire[offset:])
		offset += 8
		shareLen := int(binary.LittleEndian.Uint32(wire[offset:]))
		offset += 4
		if shareLen < 0 || offset+shareLen > len(wire) {
			return nil, fmt.Errorf("sharing: unexpected end of shares wire bytes reading entry %d payload", i)
		}
		shareBytes := make([]byte, shareLen)
		copy(shareBytes, wire[offset:offset+shareLen])
		offset += shareLen

		if _, exists := out[channelID]; exists {
			return nil, fmt.Errorf("sharing: duplicate channel id %d in shares wire bytes", channelID)
		}
		out[channelID] = shareBytes
	}

	if offset != len(wire) {
		return nil, fmt.Errorf("sharing: trailing bytes in shares wire bytes: offset=%d total=%d", offset, len(wire))
	}

	return out, nil
}
